File: dotnet/harness/Relay.Testing/S3Harness.cs
using Amazon.Runtime;
using Amazon.S3;
using Amazon.S3.Model;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Mvc.Testing;
using Testcontainers.Minio;

namespace Relay.Testing;

/// <summary>
/// S3 harness (MinIO). The container-vs-fake teaching point lives here: S3 *could* be
/// faked in-process, but MinIO is cheap and the bugs live in real HTTP + auth +
/// streaming. <c>Seed</c> = put objects directly; <c>Assert</c> = get objects back and
/// compare bytes; <c>Reset</c> = delete all objects (the bucket survives the suite).
/// </summary>
[System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1001:Types that own disposable fields should be disposable",
    Justification = "Harness lifecycle is Start/Stop (IDependencyHarness); Stop disposes the client.")]
public sealed class S3Harness<TProgram> : IDependencyHarness<TProgram>
    where TProgram : class
{
    public const string Bucket = "relay-attachments";

    private MinioContainer? _minio;
    private AmazonS3Client? _client;

    public void ConfigureWebHostBuilder(IWebHostBuilder builder)
    {
        builder.UseSetting("S3:ServiceUrl", Minio.GetConnectionString());
        builder.UseSetting("S3:AccessKey", Minio.GetAccessKey());
        builder.UseSetting("S3:SecretKey", Minio.GetSecretKey());
        builder.UseSetting("S3:Bucket", Bucket);
    }

    public async Task Start(WebApplicationFactory<TProgram> factory, CancellationToken cancellationToken)
    {
        _minio = new MinioBuilder(ContainerImages.Minio).Build();
        await _minio.StartAsync(cancellationToken);

        _client = new AmazonS3Client(
            new BasicAWSCredentials(_minio.GetAccessKey(), _minio.GetSecretKey()),
            new AmazonS3Config { ServiceURL = _minio.GetConnectionString(), ForcePathStyle = true });
        await _client.PutBucketAsync(Bucket, cancellationToken);
    }

    public async Task Stop(CancellationToken cancellationToken)
    {
        _client?.Dispose();
        if (_minio is not null)
        {
            await _minio.StopAsync(cancellationToken);
            await _minio.DisposeAsync();
        }
    }

    public async Task PutObject(string key, byte[] content, CancellationToken ct = default)
    {
        using var stream = new MemoryStream(content);
        await Client.PutObjectAsync(new PutObjectRequest
        {
            BucketName = Bucket,
            Key = key,
            InputStream = stream,
        }, ct);
    }

    public async Task<byte[]> GetObjectBytes(string key, CancellationToken ct = default)
    {
        using var response = await Client.GetObjectAsync(Bucket, key, ct);
        using var buffer = new MemoryStream();
        await response.ResponseStream.CopyToAsync(buffer, ct);
        return buffer.ToArray();
    }

    public async Task<bool> ObjectExists(string key, CancellationToken ct = default)
    {
        try
        {
            await Client.GetObjectMetadataAsync(Bucket, key, ct);
            return true;
        }
        catch (AmazonS3Exception ex) when (ex.StatusCode == System.Net.HttpStatusCode.NotFound)
        {
            return false;
        }
    }

    public async Task DeleteAllObjects(CancellationToken ct = default)
    {
        var listed = await Client.ListObjectsV2Async(new ListObjectsV2Request { BucketName = Bucket }, ct);
        if (listed.S3Objects is not { Count: > 0 })
        {
            return;
        }

        await Client.DeleteObjectsAsync(new DeleteObjectsRequest
        {
            BucketName = Bucket,
            Objects = listed.S3Objects.Select(o => new KeyVersion { Key = o.Key }).ToList(),
        }, ct);
    }

    private MinioContainer Minio => _minio ?? throw new InvalidOperationException("S3Harness is not started.");

    private AmazonS3Client Client => _client ?? throw new InvalidOperationException("S3Harness is not started.");
}

File: java/src/test/java/dev/drim/relay/harness/S3Harness.java
package dev.drim.relay.harness;

import dev.drim.relay.infra.S3Store;
import java.net.URI;
import org.testcontainers.containers.MinIOContainer;
import org.testcontainers.utility.DockerImageName;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.S3Configuration;
import software.amazon.awssdk.services.s3.model.CreateBucketRequest;
import software.amazon.awssdk.services.s3.model.Delete;
import software.amazon.awssdk.services.s3.model.DeleteObjectsRequest;
import software.amazon.awssdk.services.s3.model.GetObjectRequest;
import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
import software.amazon.awssdk.services.s3.model.NoSuchBucketException;
import software.amazon.awssdk.services.s3.model.ObjectIdentifier;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;

/**
 * The MinIO harness (object storage over real HTTP, AWS S3 SDK v2). Real container; the bucket is
 * created at boot. Seed = put objects directly; Assert = get/list + compare bytes; Reset = delete
 * all objects (the bucket survives). The teaching point: S3 *could* be faked, but real
 * HTTP+auth+streaming is where bugs live (the container-vs-fake decision). Mirrors
 * go/harness/s3.go.
 */
public final class S3Harness implements DependencyHarness {
  private final MinIOContainer container =
      new MinIOContainer(
          DockerImageName.parse(HarnessImages.MINIO).asCompatibleSubstituteFor("minio/minio"));

  private S3Client client;

  @Override
  public void start() {
    container.start();
    client =
        S3Client.builder()
            .endpointOverride(URI.create(endpoint()))
            .region(Region.US_EAST_1)
            .credentialsProvider(
                StaticCredentialsProvider.create(
                    AwsBasicCredentials.create(accessKey(), secretKey())))
            .serviceConfiguration(S3Configuration.builder().pathStyleAccessEnabled(true).build())
            .build();
    try {
      client.headBucket(b -> b.bucket(S3Store.BUCKET));
    } catch (NoSuchBucketException e) {
      client.createBucket(CreateBucketRequest.builder().bucket(S3Store.BUCKET).build());
    } catch (software.amazon.awssdk.services.s3.model.S3Exception e) {
      if (e.statusCode() == 404) {
        client.createBucket(CreateBucketRequest.builder().bucket(S3Store.BUCKET).build());
      } else {
        throw e;
      }
    }
  }

  public String endpoint() {
    return container.getS3URL();
  }

  public String accessKey() {
    return container.getUserName();
  }

  public String secretKey() {
    return container.getPassword();
  }

  @Override
  public void reset() {
    var objects =
        client
            .listObjectsV2(ListObjectsV2Request.builder().bucket(S3Store.BUCKET).build())
            .contents();
    if (objects.isEmpty()) {
      return;
    }
    var ids = objects.stream().map(o -> ObjectIdentifier.builder().key(o.key()).build()).toList();
    client.deleteObjects(
        DeleteObjectsRequest.builder()
            .bucket(S3Store.BUCKET)
            .delete(Delete.builder().objects(ids).build())
            .build());
  }

  @Override
  public void stop() {
    if (client != null) {
      client.close();
    }
    container.stop();
  }

  /** Stores an object directly — a seed. */
  public void putObject(String key, byte[] data) {
    client.putObject(
        PutObjectRequest.builder().bucket(S3Store.BUCKET).key(key).build(),
        RequestBody.fromBytes(data));
  }

  /** Reads a stored object — an Assert that bytes landed under the storage key. */
  public byte[] objectBytes(String key) {
    return client
        .getObjectAsBytes(GetObjectRequest.builder().bucket(S3Store.BUCKET).key(key).build())
        .asByteArray();
  }
}
